### PR TITLE
filter() allow named input when not calls

### LINKED ISSUE #4612: Should `filter()` allow named constants?
That'd be useful for this use case: https://stackoverflow.com/a/58798578/1725177

Maybe the name-check should only be made for symbolic inputs, that would solve the issue for both `!!!` and auto-splicing.

### PR DESCRIPTION
**Pull request by romainfrancois**
closes #4612 

``` r
library(dplyr, warn.conflicts = FALSE)

filters <- mtcars %>%
  transmute(
    cyl %in% 6:8,
    hp / drat > 50
  )

mtcars %>% filter(!!!filters)
#>     mpg cyl  disp  hp drat    wt  qsec vs am gear carb
#> 1  18.7   8 360.0 175 3.15 3.440 17.02  0  0    3    2
#> 2  14.3   8 360.0 245 3.21 3.570 15.84  0  0    3    4
#> 3  16.4   8 275.8 180 3.07 4.070 17.40  0  0    3    3
#> 4  17.3   8 275.8 180 3.07 3.730 17.60  0  0    3    3
#> 5  15.2   8 275.8 180 3.07 3.780 18.00  0  0    3    3
#> 6  10.4   8 472.0 205 2.93 5.250 17.98  0  0    3    4
#> 7  10.4   8 460.0 215 3.00 5.424 17.82  0  0    3    4
#> 8  14.7   8 440.0 230 3.23 5.345 17.42  0  0    3    4
#> 9  15.5   8 318.0 150 2.76 3.520 16.87  0  0    3    2
#> 10 13.3   8 350.0 245 3.73 3.840 15.41  0  0    3    4
#> 11 19.2   8 400.0 175 3.08 3.845 17.05  0  0    3    2
#> 12 15.8   8 351.0 264 4.22 3.170 14.50  0  1    5    4
#> 13 15.0   8 301.0 335 3.54 3.570 14.60  0  1    5    8
```

<sup>Created on 2019-12-16 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

however : 

``` r
library(dplyr, warn.conflicts = FALSE)

filter(mtcars, x = 1)
#> Error in mask$eval_all_filter(quo): filter() expressions should return logical vectors of the same size as the group
```

<sup>Created on 2019-12-16 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

### CODE REVIEW

**Comment by hadley** [comment]
I think this is probably a reasonable tradeoff, but do you mind updating the test to use `verify_output()` to make it a bit easier to see what the change is?  We also need to review the error messages for consistency.

**Comment by romainfrancois** [comment]
This only lets slide constants that are `logical` now: 

``` r
library(dplyr, warn.conflicts = FALSE)

# ok
filters <- mtcars %>%
  transmute(
    cyl %in% 6:8,
    hp / drat > 50
  )
mtcars %>% filter(!!!filters)
#>     mpg cyl  disp  hp drat    wt  qsec vs am gear carb
#> 1  18.7   8 360.0 175 3.15 3.440 17.02  0  0    3    2
#> 2  14.3   8 360.0 245 3.21 3.570 15.84  0  0    3    4
#> 3  16.4   8 275.8 180 3.07 4.070 17.40  0  0    3    3
#> 4  17.3   8 275.8 180 3.07 3.730 17.60  0  0    3    3
#> 5  15.2   8 275.8 180 3.07 3.780 18.00  0  0    3    3
#> 6  10.4   8 472.0 205 2.93 5.250 17.98  0  0    3    4
#> 7  10.4   8 460.0 215 3.00 5.424 17.82  0  0    3    4
#> 8  14.7   8 440.0 230 3.23 5.345 17.42  0  0    3    4
#> 9  15.5   8 318.0 150 2.76 3.520 16.87  0  0    3    2
#> 10 13.3   8 350.0 245 3.73 3.840 15.41  0  0    3    4
#> 11 19.2   8 400.0 175 3.08 3.845 17.05  0  0    3    2
#> 12 15.8   8 351.0 264 4.22 3.170 14.50  0  1    5    4
#> 13 15.0   8 301.0 335 3.54 3.570 14.60  0  1    5    8
  
# not ok
filter(mtcars, x = 1)
#> Error: `filter()` argument `..1` is named
#> ℹ This usually means `=` is used insted of `==`
#> ℹ Did you mean: `x == 1`
filter(mtcars, x = "A")
#> Error: `filter()` argument `..1` is named
#> ℹ This usually means `=` is used insted of `==`
#> ℹ Did you mean: `x == "A"`
filter(mtcars, x = 1 & y > 2)
#> Error: `filter()` argument `..1` is named
#> ℹ This usually means `=` is used insted of `==`
#> ℹ Did you mean: `x == 1 & y > 2`
filter(mtcars, x = 1, y > 2, z = 3)
#> Error: `filter()` argument `..1` is named
#> ℹ This usually means `=` is used insted of `==`
#> ℹ Did you mean: `x == 1`
```

<sup>Created on 2020-01-04 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

File: R/conditions.R [hunk 1]
```diff
@@ -87,3 +87,25 @@ stop_filter_eval_tidy <- function(e, index_expression) {
...
+
+#' @export
+cnd_header.dplyr_error_filter_named <- function(cnd) {
+  glue("`filter()` argument `..{cnd$index_expression}` is named")
+}
+
+#' @export
+cnd_body.dplyr_error_filter_named <- function(cnd) {
+  bullets <- c(
+    i = "This usually means `=` is used insted of `==`",
```
Comment (hadley) [inline R/conditions.R]:
```suggestion
    i = "This usually means that you've used `=` instead of `==`",
```

File: R/conditions.R [hunk 2]
```diff
@@ -87,3 +87,25 @@ stop_filter_eval_tidy <- function(e, index_expression) {
...
+    i = glue("Did you mean: `{names(cnd$quo)} == {as_label(cnd$quo[[1]])}`")
```
Comment (hadley) [inline R/conditions.R]:
```suggestion
    i = glue("Did you mean `{names(cnd$quo)} == {as_label(cnd$quo[[1]])}` ?")
```

Comment (hadley) [reply R/conditions.R]: You might also want to use `glue_data(cnd, `... {quo[[1]]}) ...`)`

File: R/tbl-df.r [hunk 3]
```diff
@@ -157,12 +157,23 @@ regroup <- function(data) {
...
-  if (any(have_name(dots))) {
-    bad <- dots[have_name(dots)]
-    bad_eq_ops(bad, "Filter specifications must not be named")
-  } else if (is_empty(dots)) {
+  if (is_empty(dots)) {
     return(.data)
   }
+  if (any(have_name(dots))) {
+    named <- have_name(dots)
+    if (any(named)) {
```
Comment (hadley) [inline R/tbl-df.r]: I think you've accidentally double up the check here

File: R/tbl-df.r [hunk 4]
```diff
@@ -157,12 +157,23 @@ regroup <- function(data) {
...
+      for (i in seq_along(named)) {
```
Comment (hadley) [inline R/tbl-df.r]: `for (i in which(named))` would things a bit, I think.

File: tests/testthat/test-filter.r [hunk 5]
```diff
@@ -445,5 +446,11 @@ test_that("filter() gives useful error messages", {
     mtcars %>%
       group_by(cyl) %>%
       filter(`_x`)
+
+    "named inputs"
+    filter(mtcars, x = 1)
+    filter(mtcars, x = "A")
```
Comment (hadley) [inline tests/testthat/test-filter.r]: I think you can probably remove this and the following test now.

I'd also test `filter(mtcars, TRUE, x = 1)`

File: R/tbl-df.r [hunk 6]
```diff
@@ -157,12 +157,23 @@ regroup <- function(data) {
...
+        quo <- dots[[i]]
+        if(named[i]) {
+          # let logical vectors slide
+          if (quo_is_call(quo) || !is.logical(quo_get_expr(quo))) {
```
Comment (hadley) [inline R/tbl-df.r]: I'm having a hard time following the logic here. Shouldn't it just be `!is.logical(quo_get_expr(quo))`?

File: R/tbl-df.r [hunk 7]
```diff
@@ -157,12 +157,23 @@ regroup <- function(data) {
 #' @export
 filter.tbl_df <- function(.data, ..., .preserve = FALSE) {
   dots <- enquos(...)
-  if (any(have_name(dots))) {
-    bad <- dots[have_name(dots)]
-    bad_eq_ops(bad, "Filter specifications must not be named")
```
Comment (hadley) [inline R/tbl-df.r]: I suspect you can now remove `bad_eq_ops()`?

File: tests/testthat/test-filter.r [hunk 8]
```diff
@@ -445,5 +446,11 @@ test_that("filter() gives useful error messages", {
     mtcars %>%
       group_by(cyl) %>%
       filter(`_x`)
+
+    "named inputs"
```
Comment (hadley) [inline tests/testthat/test-filter.r]: BTW if you do `"# named inputs"` it'll become a heading in the output.

**Comment by hadley** [comment]
BTW after this is merged we should probably do #4682, since it won't cause merge conflicts with any open PRs

File: R/tbl-df.r (hunk 6 above)
Comment (romainfrancois) [reply R/tbl-df.r]: Yeah, thanks. No need to separate calls from non calls

File: tests/testthat/test-filter.r (hunk 8 above)
Comment (romainfrancois) [reply tests/testthat/test-filter.r]: Nice thanks.

File: R/tbl-df.r (hunk 7 above)
Comment (romainfrancois) [reply R/tbl-df.r]: ✅

**Comment by romainfrancois** [comment]
Deal, especially if we start tracking errors with `verify_output()`.

**Pull request merged**